src/v/redpanda/admin/security.cc

```diff
@@ -476,3 +496,58 @@ admin_server::oidc_revoke_handler(std::unique_ptr<ss::http::request>) {
     });
     co_return ss::json::json_return_type(ss::json::json_void());
 }
+
+ss::future<ss::json::json_return_type>
+admin_server::create_role_handler(std::unique_ptr<ss::http::request> req) {
+    if (need_redirect_to_leader(model::controller_ntp, _metadata_cache)) {
+        // In order that we can do a reliably ordered validation of
+        // the request (and drop no-op requests), run on controller leader;
+        throw co_await redirect_to_leader(*req, model::controller_ntp);
+    }
+
+    auto doc = co_await parse_json_body(req.get());
+
+    if (!doc.IsObject()) {
+        vlog(adminlog.debug, "Request body is not a JSON object");
+        throw_role_exception(
+          role_errc::malformed_def, "Request body is not a JSON object");
+    }
+
+    if (!doc.HasMember("role") || !doc["role"].IsString()) {
+        vlog(adminlog.debug, "String 'role' missing from request body");
+        throw_role_exception(
+          role_errc::malformed_def, "Missing string field 'role'");
+    }
+
+    auto role_name = security::role_name{doc["role"].GetString()};
+    validate_no_control(role_name(), string_conversion_exception{role_name()});
+
+    if (!security::validate_scram_username(role_name())) {
+        throw_role_exception(role_errc::invalid_name);
+    }
+
+    ss::httpd::security_json::role_definition j_res;
+    j_res.role = role_name();
+
+    if (is_no_op_role_write(
+          _controller->get_role_store().local(), role_name, security::role{})) {
+        vlog(adminlog.debug, "Empty role {} already exists", role_name);
+        co_return ss::json::json_return_type(j_res);
+    }
+
+    auto err
+      = co_await _controller->get_security_frontend().local().create_role(
+        role_name, security::role{}, model::timeout_clock::now() + 5s);
+
+    if (err == cluster::errc::role_exists) {
+        // Idempotency: if the empty role already exists,
+        // suppress the role_exists error and return success.
+        auto role = _controller->get_role_store().local().get(role_name);
+        if (role.has_value() && role.value().members().empty()) {
+            co_return ss::json::json_return_type(j_res);
+        }
+    }
+
+    co_await throw_on_error(*req, err, model::controller_ntp);
+    co_return ss::json::json_return_type(j_res);
```

Comment:
Something else I've been meaning to address - we intended to return 201 (Created) on this endpoint because REST... I don't know how easy that is. It's possible that the seastar machinery is clever enough to glean it from the swagger, but I'm not sure.

Reply:
I don't think it's smart enough, unfortunately. It's returning 200 (OK) at the moment.

I spent a bit of time trying to refactor this to allow setting a result code, but it's not straightforward as far as I can tell. We'd have to at least refactor `register_route` and replicate what seastar is doing for creating a json handler for this to work.

I'd be tempted to keep this returning 200 to be consistent with the other admin API endpoints (and to keep things simple), but let me know what you think.

Reply:
Bummer. I'll add it my short list of minor API changes. Nothing for you to do, but I'll post an update on the channel today. I have a couple others, also arising from seastar::httpd implementation details

Reply:
It can be done, but you have to be a bit more manual about it: https://github.com/redpanda-data/redpanda/blob/dev/src/v/pandaproxy/reply.h#L75-L82

Reply:
Nice. We'd need access to the http_reply object, so we'd want to use [register_route_raw_async](https://github.com/redpanda-data/redpanda/blob/8ee2339cf5b16deae9a0323794c5b3fbc4c0febe/src/v/redpanda/admin/server.h#L293-L295) to register the route IIUC

Reply:
My mistake, I think this is the correct overload:

https://github.com/redpanda-data/redpanda/blob/dc7b8d8a4e20ffaf201052c5a7714cece0119735/src/v/redpanda/admin/server.h#L392-L398

Reply:
Created an issue to make this change as a follow up: https://github.com/redpanda-data/core-internal/issues/1173